airbyte-cdk/python/airbyte_cdk/sources/streams/utils/stream_helper.py

```diff
@@ -34,6 +35,11 @@ def get_first_record_for_slice(stream: Stream, stream_slice: Optional[Mapping[st
     :raises StopIteration: if there is no first record to return (the read_records generator is empty)
     :return: StreamData containing the first record in the slice
     """
+    if isinstance(stream, DeclarativeStream):
```

Comment:
This code does not respect the Liskov substitution principle. Can we just call `stream.exit_on_rate_limit` on a DeclarativeStream?

Also, should the method `get_first_record_for_slice` be removed following https://github.com/airbytehq/airbyte/pull/40682? From a quick look, the usage I see of this method is only in the availability strategy and in stripe (for its own availability strategy)

Reply:
> This code does not respect the Liskov substitution principle. Can we just call stream.exit_on_rate_limit on a DeclarativeStream?

I partially agree with you. Having one attribute in different places can be confusing. However, the issue is that for declarative behavior, we want to isolate the conduction of components. exit_on_rate_limit is obviously something related to requests and should be handled there. If we add it to stream, it will break the architectural logic, and we will end up in a situation where both the stream and the retriever operate with the parameter of the request.

On the other hand, we do specify that we expect the stream to be of the Stream type, but in reality, this is only used in HttpStream and DeclarativeStream.

> Also, should the method get_first_record_for_slice be removed following https://github.com/airbytehq/airbyte/pull/40682? From a quick look, the usage I see of this method is only in the availability strategy and in stripe (for its own availability strategy)

I'm not sure, as we use it in the availability strategy, which is used in the CheckStream component of low code.

Reply:
> On the other hand, we do specify that we expect the stream to be of the Stream type, but in reality, this is only used in HttpStream and DeclarativeStream.

This means that this is even more dangerous than I thought as this code could call `stream.exit_on_rate_limit` on a `airbyte_cdk.sources.streams.core.Stream` implementation that is not DeclarativeStream nor HttpStream and this would generate an `AttributeError`. This is the danger of not respecting the Liskov principle and it means that this will break for `Describe` in Salesforce for example

Reply:
I have update approach to safety set `exit_on_rate_limit`.